Enums cannot be used as tuple elements

Consider the following code:

```
o: output sum of int;
type Foo = enum {
    BAR = "BAR",
    BAZ = "BAZ"
};

type Quux = { id: int, isa: Foo };
o << 1;
```

This gives the error:
```
test-enum/boa/Test.java:137: error: enum types may not be instantiated
                                this.___isa = new BoaEnum_0(___isa);
```

This error prevents the use of enum types as elements of tuples.

--- a/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
+++ b/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
@@ -320,7 +320,7 @@ public class CodeGeneratingVisitor extends AbstractCodeGeneratingVisitor {
 				}
 				fieldCount++;
 				BoaType type = c.getType().type;
-				protos.add(type instanceof BoaProtoTuple);
+				protos.add((type instanceof BoaProtoTuple) || (type instanceof BoaEnum));
 				types.add(type.toBoxedJavaType());
 			}
 

--- a/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
+++ b/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
@@ -309,7 +309,7 @@ public class CodeGeneratingVisitor extends AbstractCodeGeneratingVisitor {
 			final List<Component> members = n.getMembers();
 			final List<String> fields = new ArrayList<String>();
 			final List<String> types = new ArrayList<String>();
-			final List<Boolean> nocopies = new ArrayList<Boolean>();
+			final List<Boolean> aliases = new ArrayList<Boolean>();
 
 			int fieldCount = 0;
 			for (final Component c : members) {
@@ -320,14 +320,14 @@ public class CodeGeneratingVisitor extends AbstractCodeGeneratingVisitor {
 				}
 				fieldCount++;
 				BoaType type = c.getType().type;
-				nocopies.add((type instanceof BoaProtoTuple) || (type instanceof BoaEnum));
+				aliases.add((type instanceof BoaProtoTuple) || (type instanceof BoaEnum));
 				types.add(type.toBoxedJavaType());
 			}
 
 			st.add("name", tupType.toJavaType());
 			st.add("fields", fields);
 			st.add("types", types);
-			st.add("nocopies", nocopies);
+			st.add("aliases", aliases);
 
 			code.add(st.render());
 		}

--- a/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
+++ b/src/java/boa/compiler/visitors/CodeGeneratingVisitor.java
@@ -309,7 +309,7 @@ public class CodeGeneratingVisitor extends AbstractCodeGeneratingVisitor {
 			final List<Component> members = n.getMembers();
 			final List<String> fields = new ArrayList<String>();
 			final List<String> types = new ArrayList<String>();
-			final List<Boolean> protos = new ArrayList<Boolean>();
+			final List<Boolean> nocopies = new ArrayList<Boolean>();
 
 			int fieldCount = 0;
 			for (final Component c : members) {
@@ -320,14 +320,14 @@ public class CodeGeneratingVisitor extends AbstractCodeGeneratingVisitor {
 				}
 				fieldCount++;
 				BoaType type = c.getType().type;
-				protos.add((type instanceof BoaProtoTuple) || (type instanceof BoaEnum));
+				nocopies.add((type instanceof BoaProtoTuple) || (type instanceof BoaEnum));
 				types.add(type.toBoxedJavaType());
 			}
 
 			st.add("name", tupType.toJavaType());
 			st.add("fields", fields);
 			st.add("types", types);
-			st.add("protos", protos);
+			st.add("nocopies", nocopies);
 
 			code.add(st.render());
 		}

--- a/templates/BoaJava.stg
+++ b/templates/BoaJava.stg
@@ -125,16 +125,16 @@ new <type>()
 }
 >>
 
-TupleType(name, fields, types, nocopies) ::= <<
+TupleType(name, fields, types, aliases) ::= <<
 private class <name>
 {
 	<fields,types:{field,type|<type> ___<field>;}; separator="\n">
 
 	<name>(<fields,types:{field,type|<type> ___<field>}; separator=", ">) {
-		<fields,types,nocopies:{field,type,nocopy|<if(nocopy)>this.___<field> = ___<field>;<else>this.___<field> = new <type>(___<field>);<endif>}; separator="\n">
+		<fields,types,aliases:{field,type,alias|<if(alias)>this.___<field> = ___<field>;<else>this.___<field> = new <type>(___<field>);<endif>}; separator="\n">
 	}
 	<name>(<name> tmp) {
-		<fields,types,nocopies:{field,type,nocopy|<if(nocopy)> this.___<field> = tmp.___<field>;<else> this.___<field> = new <type>(tmp.___<field>);<endif>}; separator="\n">
+		<fields,types,aliases:{field,type,alias|<if(alias)> this.___<field> = tmp.___<field>;<else> this.___<field> = new <type>(tmp.___<field>);<endif>}; separator="\n">
 	}
 
 	public <name> clone() {

--- a/templates/BoaJava.stg
+++ b/templates/BoaJava.stg
@@ -125,16 +125,16 @@ new <type>()
 }
 >>
 
-TupleType(name, fields, types, protos) ::= <<
+TupleType(name, fields, types, nocopies) ::= <<
 private class <name>
 {
 	<fields,types:{field,type|<type> ___<field>;}; separator="\n">
 
 	<name>(<fields,types:{field,type|<type> ___<field>}; separator=", ">) {
-		<fields,types,protos:{field,type,proto|<if(proto)>this.___<field> = ___<field>;<else>this.___<field> = new <type>(___<field>);<endif>}; separator="\n">
+		<fields,types,nocopies:{field,type,nocopy|<if(nocopy)>this.___<field> = ___<field>;<else>this.___<field> = new <type>(___<field>);<endif>}; separator="\n">
 	}
 	<name>(<name> tmp) {
-		<fields,types,protos:{field,type,proto|<if(proto)> this.___<field> = tmp.___<field>;<else> this.___<field> = new <type>(tmp.___<field>);<endif>}; separator="\n">
+		<fields,types,nocopies:{field,type,nocopy|<if(nocopy)> this.___<field> = tmp.___<field>;<else> this.___<field> = new <type>(tmp.___<field>);<endif>}; separator="\n">
 	}
 
 	public <name> clone() {

--- a/test/codegen/issue285.boa
+++ b/test/codegen/issue285.boa
@@ -1,0 +1,8 @@
+o: output sum of int;
+type Foo = enum {
+    BAR = "BAR",
+    BAZ = "BAZ"
+};
+
+type Quux = { id: int, isa: Foo };
+o << 1;

